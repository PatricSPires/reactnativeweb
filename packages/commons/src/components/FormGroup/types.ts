type Input = {
  name?: string
  value?: string
  onFocus?: () => void
  onBlur: () => void
}

export type FormGroupProps = {
  name: string
  tag: unknown
  label?: string
  width?: string
  labelHint?: string
  activeAppearance?: string
  className?: string
  placeholder?: string
  helperText?: string
  meta?: {
    active?: boolean
    touched?: boolean
    error?: string
    valid?: boolean
  }
  input?: Input
  iconPosition?: 'left' | 'right'
  icon?: React.ReactNode
  iconColor?: 'error' | 'warning' | 'success'
  isInvalid?: boolean
  passwordField?: boolean
  marginBottom: 1 | 2 | 3 | 4 | 5 | 6 | 7
}

export type FormGroupWrapperProps = Pick<
  FormGroupProps,
  'width' | 'marginBottom'
>
