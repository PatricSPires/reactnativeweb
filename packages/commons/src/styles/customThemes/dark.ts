import { DefaultTheme } from 'styled-components'
import { gray, white } from '../colors'
import theme from '../theme'

const dark: DefaultTheme = {
  ...theme,
  title: 'dark',

  colors: {
    ...theme.colors,
    primary_container_background: gray[900],
    medium_contrast_border: gray[700],
    high_contrast_border: gray[500],
    font_color: gray[500],
    primary_color_light: gray[500],
  },
  supportColors: {
    ...theme.supportColors,
    information: white[500],
  },
}

export default dark
