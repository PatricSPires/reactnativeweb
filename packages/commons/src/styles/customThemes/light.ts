import { DefaultTheme } from 'styled-components'
import theme from '../theme'
import { white, gray, blue } from '../colors'

const light: DefaultTheme = {
  ...theme,
  title: 'light',

  colors: {
    primary_container_background: white[100],
    medium_contrast_border: gray[700],
    primary_color_light: blue[700],
    high_contrast_border: gray[500],
    font_color: gray[500],
  },
  supportColors: {
    information: white[500],
  },
}

export default light
