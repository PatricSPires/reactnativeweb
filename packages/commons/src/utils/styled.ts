import styleSheetSerializer from 'jest-styled-components/src/styleSheetSerializer'

export function enableStyledComponentsSerializer() {
  expect.addSnapshotSerializer(styleSheetSerializer)
}
