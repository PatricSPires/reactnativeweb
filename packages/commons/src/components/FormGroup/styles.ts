import styled from 'styled-components'
import { View, Text } from 'react-native'
import theme from '../../styles/theme'
import { FormGroupWrapperProps } from './types'

export const FormGroupWrapper = styled(View)<FormGroupWrapperProps>`
  display: flex;
  width: ${({ width }) => width || '100%'};
  margin-bottom: ${({ marginBottom }) => theme.spacings[marginBottom || 3]};

  Tag {
    margin: 8px 0 4px 0;
  }
`

export const Label = styled(Text)`
  color: ${(props) => props.theme.colors.font_color};
  font-size: ${theme.fontSize.label_01};
  margin-bottom: ${theme.spacings[1]};
  display: flex;
  align-items: center;
`

export const HelperText = styled(Text)`
  font-size: ${theme.fontSize.label_01};
  color: ${theme.colors.medium_contrast_border};
`

export const ErrorMessage = styled(Text)`
  font-size: ${theme.fontSize.label_01};
  color: ${theme.supportColors.error};
  margin-top: ${theme.spacings[1]};
`
