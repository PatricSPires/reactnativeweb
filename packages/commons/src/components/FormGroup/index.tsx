// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-nocheck
// TODO: Tentamos achar uma forma de corrigir a tipagem da tag, mas, infelizmente falhamos, a final somos humanos.
import React from 'react'
import { FormGroupProps } from './types'
import { FormGroupWrapper, Label, HelperText, ErrorMessage } from './styles'
import Tooltip from '../Tooltip'

const FormGroup = ({
  tag: Tag,
  label,
  width,
  labelHint,
  placeholder,
  helperText,
  iconPosition,
  icon,
  iconColor,
  passwordField,
  marginBottom,
  meta = {},
  input = {
    onBlur: Function,
  },
  ...props
}: FormGroupProps) => {
  const isActive = meta.active
  const isInvalid = meta.error !== undefined && meta.touched && !isActive
  const isValid = meta.valid && meta.touched && !isActive

  return (
    <FormGroupWrapper width={width} marginBottom={marginBottom}>
      {!!label && (
        <Label>
          {label}
          {/* {!!labelHint && <Tooltip hint={labelHint} />} */}
        </Label>
      )}
      {!!helperText && <HelperText>{helperText}</HelperText>}
      <Tag
        {...input}
        {...props}
        placeholder={placeholder}
        active={isActive}
        isInvalid={isInvalid}
        isValid={isValid}
        iconPosition={iconPosition}
        icon={icon}
        iconColor={iconColor}
        passwordField={passwordField}
      />
      {!!isInvalid && <ErrorMessage>{meta.error}</ErrorMessage>}
    </FormGroupWrapper>
  )
}

FormGroup.defaultProps = {
  marginBottom: 3,
}

export default FormGroup
