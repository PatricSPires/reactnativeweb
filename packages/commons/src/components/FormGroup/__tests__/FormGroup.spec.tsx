import * as React from 'react'
import { render } from '@testing-library/react-native'
import { ThemeProvider } from 'styled-components'
import { enableStyledComponentsSerializer } from '../../../utils/styled'
import FormGroup from '../index'

import dark from '../../../styles/customThemes/dark'
import light from '../../../styles/customThemes/light'

const props = {
  tag: 'input',
  name: 'input',
  meta: {
    error: 'Error message',
    touched: true,
    active: false,
  },
}

beforeAll(() => {
  enableStyledComponentsSerializer()
})

describe('FormGroup Test', () => {
  it('should render the FormGroup', () => {
    const { toJSON } = render(
      <FormGroup {...props} name="FormGroup" tag="input" />,
    )

    expect(toJSON()).toMatchSnapshot()
  })

  it('should render the FormGroup with label and dark theme', () => {
    const { toJSON } = render(
      <ThemeProvider theme={dark}>
        <FormGroup
          {...props}
          label="Hello World"
          name="FormGroup"
          tag="input"
        />
      </ThemeProvider>,
    )

    expect(toJSON()).toMatchSnapshot()
  })

  it('should render the FormGroup with label light theme', () => {
    const { toJSON } = render(
      <ThemeProvider theme={light}>
        <FormGroup
          {...props}
          label="Hello World"
          name="FormGroup"
          tag="input"
        />
      </ThemeProvider>,
    )

    expect(toJSON()).toMatchSnapshot()
  })

  it('should render the FormGroup with placeholder', () => {
    const { toJSON } = render(
      <FormGroup
        {...props}
        placeholder="Hello World"
        name="FormGroup"
        tag="input"
      />,
    )

    expect(toJSON()).toMatchSnapshot()
  })

  it('should render the FormGroup with helper text', () => {
    const { toJSON } = render(
      <FormGroup
        {...props}
        helperText="Hello World"
        name="FormGroup"
        tag="input"
      />,
    )

    expect(toJSON()).toMatchSnapshot()
  })

  it('should render the FormGroup with isInvalid true', () => {
    const { toJSON } = render(
      <FormGroup {...props} name="FormGroup" tag="input" />,
    )

    expect(toJSON()).toMatchSnapshot()
  })

  it('should render the FormGroup with width 100%', () => {
    const { toJSON } = render(
      <FormGroup {...props} width="100%" name="FormGroup" tag="input" />,
    )

    expect(toJSON()).toMatchSnapshot()
  })

  it('should render the FormGroup with isInvalid false', () => {
    const customProps = {
      ...props,
      meta: {
        ...props.meta,
        error: undefined,
      },
    }

    const { toJSON } = render(
      <FormGroup {...customProps} name="FormGroup" tag="input" />,
    )
    expect(toJSON()).toMatchSnapshot()
  })

  it('should render the FormGroup with isValid true', () => {
    const customProps = {
      ...props,
      meta: {
        ...props.meta,
        valid: true,
      },
    }

    const { toJSON } = render(
      <FormGroup {...customProps} name="FormGroup" tag="input" />,
    )
    expect(toJSON()).toMatchSnapshot()
  })

  it('should render the FormGroup with labelHint with dark theme', () => {
    const { toJSON } = render(
      <ThemeProvider theme={dark}>
        <FormGroup
          {...props}
          label="Hello World"
          labelHint="Helo FormGroup"
          name="FormGroup"
          tag="input"
        />
      </ThemeProvider>,
    )
    expect(toJSON()).toMatchSnapshot()
  })

  it('should render the FormGroup with labelHint with light theme', () => {
    const { toJSON } = render(
      <ThemeProvider theme={light}>
        <FormGroup
          {...props}
          label="Hello World"
          labelHint="Helo FormGroup"
          name="FormGroup"
          tag="input"
        />
      </ThemeProvider>,
    )
    expect(toJSON()).toMatchSnapshot()
  })
})
