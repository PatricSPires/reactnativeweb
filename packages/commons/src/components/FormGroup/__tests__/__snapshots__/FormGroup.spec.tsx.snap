// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormGroup Test should render the FormGroup 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  margin-bottom: 24px;
}

.c0 Tag {
  margin: 8px 0 4px 0;
}

.c1 {
  font-size: .75rem;
  color: #DA1E28;
  margin-top: 8px;
}

<div
  className="c0 sc-bdnylx c0 css-view-1dbjc4n"
>
  <input
    active={false}
    isInvalid={true}
    name="FormGroup"
    onBlur={[Function]}
  />
  <div
    className="c1 sc-hKFyIo c1 css-text-901oao"
    dir="auto"
  >
    Error message
  </div>
</div>
`;

exports[`FormGroup Test should render the FormGroup with helper text 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  margin-bottom: 24px;
}

.c0 Tag {
  margin: 8px 0 4px 0;
}

.c1 {
  font-size: .75rem;
  color: #8d8d8d;
}

.c2 {
  font-size: .75rem;
  color: #DA1E28;
  margin-top: 8px;
}

<div
  className="c0 sc-bdnylx c0 css-view-1dbjc4n"
>
  <div
    className="c1 sc-dlnjPT c1 css-text-901oao"
    dir="auto"
  >
    Hello World
  </div>
  <input
    active={false}
    isInvalid={true}
    name="FormGroup"
    onBlur={[Function]}
  />
  <div
    className="c2 sc-hKFyIo c2 css-text-901oao"
    dir="auto"
  >
    Error message
  </div>
</div>
`;

exports[`FormGroup Test should render the FormGroup with isInvalid false 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  margin-bottom: 24px;
}

.c0 Tag {
  margin: 8px 0 4px 0;
}

<div
  className="c0 sc-bdnylx c0 css-view-1dbjc4n"
>
  <input
    active={false}
    isInvalid={false}
    name="FormGroup"
    onBlur={[Function]}
  />
</div>
`;

exports[`FormGroup Test should render the FormGroup with isInvalid true 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  margin-bottom: 24px;
}

.c0 Tag {
  margin: 8px 0 4px 0;
}

.c1 {
  font-size: .75rem;
  color: #DA1E28;
  margin-top: 8px;
}

<div
  className="c0 sc-bdnylx c0 css-view-1dbjc4n"
>
  <input
    active={false}
    isInvalid={true}
    name="FormGroup"
    onBlur={[Function]}
  />
  <div
    className="c1 sc-hKFyIo c1 css-text-901oao"
    dir="auto"
  >
    Error message
  </div>
</div>
`;

exports[`FormGroup Test should render the FormGroup with isValid true 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  margin-bottom: 24px;
}

.c0 Tag {
  margin: 8px 0 4px 0;
}

.c1 {
  font-size: .75rem;
  color: #DA1E28;
  margin-top: 8px;
}

<div
  className="c0 sc-bdnylx c0 css-view-1dbjc4n"
>
  <input
    active={false}
    isInvalid={true}
    isValid={true}
    name="FormGroup"
    onBlur={[Function]}
  />
  <div
    className="c1 sc-hKFyIo c1 css-text-901oao"
    dir="auto"
  >
    Error message
  </div>
</div>
`;

exports[`FormGroup Test should render the FormGroup with label and dark theme 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  margin-bottom: 24px;
}

.c0 Tag {
  margin: 8px 0 4px 0;
}

.c1 {
  color: #919191;
  font-size: .75rem;
  margin-bottom: 8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  font-size: .75rem;
  color: #DA1E28;
  margin-top: 8px;
}

<div
  className="c0 sc-bdnylx c0 css-view-1dbjc4n"
>
  <div
    className="c1 sc-gtssRu c1 css-text-901oao"
    dir="auto"
  >
    Hello World
  </div>
  <input
    active={false}
    isInvalid={true}
    name="FormGroup"
    onBlur={[Function]}
  />
  <div
    className="c2 sc-hKFyIo c2 css-text-901oao"
    dir="auto"
  >
    Error message
  </div>
</div>
`;

exports[`FormGroup Test should render the FormGroup with label light theme 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  margin-bottom: 24px;
}

.c0 Tag {
  margin: 8px 0 4px 0;
}

.c1 {
  color: #919191;
  font-size: .75rem;
  margin-bottom: 8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  font-size: .75rem;
  color: #DA1E28;
  margin-top: 8px;
}

<div
  className="c0 sc-bdnylx c0 css-view-1dbjc4n"
>
  <div
    className="c1 sc-gtssRu c1 css-text-901oao"
    dir="auto"
  >
    Hello World
  </div>
  <input
    active={false}
    isInvalid={true}
    name="FormGroup"
    onBlur={[Function]}
  />
  <div
    className="c2 sc-hKFyIo c2 css-text-901oao"
    dir="auto"
  >
    Error message
  </div>
</div>
`;

exports[`FormGroup Test should render the FormGroup with labelHint with dark theme 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  margin-bottom: 24px;
}

.c0 Tag {
  margin: 8px 0 4px 0;
}

.c1 {
  color: #919191;
  font-size: .75rem;
  margin-bottom: 8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c6 {
  font-size: .75rem;
  color: #DA1E28;
  margin-top: 8px;
}

.c3 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c5 {
  visibility: hidden;
  width: 120px;
  background-color: #161616;
  color: #ffffff;
  text-align: center;
  border-radius: 6px;
  padding: 5px 0;
  position: absolute;
  z-index: 1;
  top: -220%;
  left: 68%;
  margin-left: -60px;
}

.c5::after {
  content: '';
  position: absolute;
  bottom: -38%;
  left: 50%;
  margin-left: -5px;
  border-width: 5px;
  border-style: solid;
  border-color: transparent transparent #161616 transparent;
  -webkit-transform: rotate(180deg);
  -ms-transform: rotate(180deg);
  transform: rotate(180deg);
}

.c2 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c2 svg {
  width: 16px;
  height: 16px;
  margin-left: 10px;
}

.c2:hover .c4 {
  visibility: visible;
}

<div
  className="c0 sc-bdnylx c0 css-view-1dbjc4n"
>
  <div
    className="c1 sc-gtssRu c1 css-text-901oao"
    dir="auto"
  >
    Hello World
    <div
      className="c2 sc-jSFkmK c2 css-view-1dbjc4n r-display-xoduu5"
    >
      <svg
        aria-hidden="true"
        className="c3"
        fill="none"
        focusable="false"
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx={12}
          cy={12}
          r={10}
        />
        <line
          x1={12}
          x2={12}
          y1={8}
          y2={12}
        />
        <line
          x1={12}
          x2={12.01}
          y1={16}
          y2={16}
        />
      </svg>
      <span
        className="c4 c5 c4 c5 css-text-901oao css-textHasAncestor-16my406"
        dir="auto"
      >
        Helo FormGroup
      </span>
    </div>
  </div>
  <input
    active={false}
    isInvalid={true}
    name="FormGroup"
    onBlur={[Function]}
  />
  <div
    className="c6 sc-hKFyIo c6 css-text-901oao"
    dir="auto"
  >
    Error message
  </div>
</div>
`;

exports[`FormGroup Test should render the FormGroup with labelHint with light theme 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  margin-bottom: 24px;
}

.c0 Tag {
  margin: 8px 0 4px 0;
}

.c1 {
  color: #919191;
  font-size: .75rem;
  margin-bottom: 8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c6 {
  font-size: .75rem;
  color: #DA1E28;
  margin-top: 8px;
}

.c3 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c5 {
  visibility: hidden;
  width: 120px;
  background-color: #161616;
  color: #ffffff;
  text-align: center;
  border-radius: 6px;
  padding: 5px 0;
  position: absolute;
  z-index: 1;
  top: -220%;
  left: 68%;
  margin-left: -60px;
}

.c5::after {
  content: '';
  position: absolute;
  bottom: -38%;
  left: 50%;
  margin-left: -5px;
  border-width: 5px;
  border-style: solid;
  border-color: transparent transparent #161616 transparent;
  -webkit-transform: rotate(180deg);
  -ms-transform: rotate(180deg);
  transform: rotate(180deg);
}

.c2 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c2 svg {
  width: 16px;
  height: 16px;
  margin-left: 10px;
}

.c2:hover .c4 {
  visibility: visible;
}

<div
  className="c0 sc-bdnylx c0 css-view-1dbjc4n"
>
  <div
    className="c1 sc-gtssRu c1 css-text-901oao"
    dir="auto"
  >
    Hello World
    <div
      className="c2 sc-jSFkmK c2 css-view-1dbjc4n r-display-xoduu5"
    >
      <svg
        aria-hidden="true"
        className="c3"
        fill="none"
        focusable="false"
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx={12}
          cy={12}
          r={10}
        />
        <line
          x1={12}
          x2={12}
          y1={8}
          y2={12}
        />
        <line
          x1={12}
          x2={12.01}
          y1={16}
          y2={16}
        />
      </svg>
      <span
        className="c4 c5 c4 c5 css-text-901oao css-textHasAncestor-16my406"
        dir="auto"
      >
        Helo FormGroup
      </span>
    </div>
  </div>
  <input
    active={false}
    isInvalid={true}
    name="FormGroup"
    onBlur={[Function]}
  />
  <div
    className="c6 sc-hKFyIo c6 css-text-901oao"
    dir="auto"
  >
    Error message
  </div>
</div>
`;

exports[`FormGroup Test should render the FormGroup with placeholder 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  margin-bottom: 24px;
}

.c0 Tag {
  margin: 8px 0 4px 0;
}

.c1 {
  font-size: .75rem;
  color: #DA1E28;
  margin-top: 8px;
}

<div
  className="c0 sc-bdnylx c0 css-view-1dbjc4n"
>
  <input
    active={false}
    isInvalid={true}
    name="FormGroup"
    onBlur={[Function]}
    placeholder="Hello World"
  />
  <div
    className="c1 sc-hKFyIo c1 css-text-901oao"
    dir="auto"
  >
    Error message
  </div>
</div>
`;

exports[`FormGroup Test should render the FormGroup with width 100% 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  margin-bottom: 24px;
}

.c0 Tag {
  margin: 8px 0 4px 0;
}

.c1 {
  font-size: .75rem;
  color: #DA1E28;
  margin-top: 8px;
}

<div
  className="c0 sc-bdnylx c0 css-view-1dbjc4n"
>
  <input
    active={false}
    isInvalid={true}
    name="FormGroup"
    onBlur={[Function]}
  />
  <div
    className="c1 sc-hKFyIo c1 css-text-901oao"
    dir="auto"
  >
    Error message
  </div>
</div>
`;
